packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,530 @@
+import {Metrics} from "../../metrics/index.js";
+import {IClock} from "../../util/clock.js";
+import {ShufflingCache} from "../shufflingCache.js";
+import {MapTracker} from "./mapMetrics.js";
+import {
+  CacheType,
+  CheckpointHex,
+  PersistentCheckpointStateCacheModules,
+  PersistentCheckpointStateCacheOpts,
+  GetHeadStateFn,
+  RemovePersistedStateReason,
+  CheckpointStateCache,
+  CheckpointKey,
+} from "./types.js";
+import {CPStatePersistentApis, PersistentKey} from "./persistent/types.js";
+
+/**
+ * An implementation of CheckpointStateCache that keep up to n epoch checkpoint states in memory and persist the rest to disk
+ * - If it's more than `maxEpochsInMemory` epochs old, it will be persisted to disk following LRU cache
+ * - Once a chain gets finalized we'll prune all states from memory and disk for epochs < finalizedEpoch
+ * - In get*() apis if shouldReload is true, it will reload from disk. The reload() api is expensive (as with Holesky, it takes ~1.5s to load and could be
+ * up 2s-3s in total for the hashTreeRoot() ) and should only be called in some important flows:
+ *   - Get state for block processing
+ *   - updateHeadState
+ *   - as with any cache, the state could be evicted from memory at any time, so we should always check if the state is in memory or not
+ * - For each epoch, we only persist exactly 1 (official) checkpoint state and prune the other one because it's enough for the regen. The persisted (official)
+ * checkpoint state could be finalized and used later in archive task. The "official" checkpoint state is defined at: https://github.com/ethereum/consensus-specs/blob/v1.4.0-beta.2/specs/phase0/beacon-chain.md
+ *
+ *   - If there is Current Root Checkpoint State, we persist that state to disk and delete the Previous Root Checkpoint State
+ *        epoch:       (n-2)   (n-1)     n     (n+1)
+ *               |-------|-------|-------|-------|
+ *        root      ---------------------^
+ *
+ *   - If there is no Current Root Checkpoint State, we persist the Previous Root Checkpoint State to disk
+ *        epoch:       (n-2)   (n-1)     n     (n+1)
+ *               |-------|-------|-------|-------|
+ *        root     ---------------------^
+ *
+ * The below diagram shows Previous Root Checkpoint State is persisted for epoch (n-2) and Current Root Checkpoint State is persisted for epoch (n-1)
+ * while at epoch (n) and (n+1) we have both of them in memory
+ *
+ * ╔════════════════════════════════════╗═══════════════╗
+ * ║      persisted to db or fs         ║   in memory   ║
+ * ║        reload if needed            ║               ║
+ * ║ -----------------------------------║---------------║
+ * ║        epoch:       (n-2)   (n-1)  ║  n     (n+1)  ║
+ * ║               |-------|-------|----║--|-------|----║
+ * ║                      ^        ^    ║ ^       ^     ║
+ * ║                                    ║  ^       ^    ║
+ * ╚════════════════════════════════════╝═══════════════╝
+ */
```

Comment:
Excellent module-wide description of the mechanism!! Super helpful an appreciated, I would encourage to do this level of documentation on more functions and new pieces that are not naive.